### DISCUSSION TITLE
Shader abilities

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3817] [Q&A] [answered]
I understand that the usual goal of a material fragment block is to compute material properties, but can it do other things permitted in OpenGL ES fragment shaders, such as `discard` the fragment or write to the `gl_FragDepth` variable?

**Comment by romainguy** [comment, marked as answer]
You can do anything GLSL support but *we* don't support writing to gl_* variables. Meaning that we may break your code in the future since we make no guarantees (esp as we will eventually move away from GL and focus on Vulkan and Metal). The use of discard is fully supported though.

**Reply by jwwalker** [reply]
If you break writing to `gl_FragDepth` I hope you will provide some alternate way of modifying the depth of the current fragment.

**Reply by romainguy** [reply]
What are you doing with this btw? Modifying gl_FragDepth is never recommended (for performance reasons).

**Reply by jwwalker** [reply]
What I'm trying to do is draw a colored outline around a selected object, and I want points of the outline to have the same depth as nearby points of the object.  Here's a simple example:

![outlined object](https://user-images.githubusercontent.com/730426/115599520-942e8880-a290-11eb-8041-9fb21ef10963.png)

In an old OpenGL program, I did that by a method that drew the object in edge mode with line thickness and using the stencil buffer to mask out the object being outlined.   But in Filament there is no edge mode and no line thickness control. My attempt to produce a similar effect in Filament involves drawing the selected object into a depth texture, then drawing a view-filling quad with a material that samples the depth texture.  If the fragment shader determines that the current fragment is in the selected object, or far away from it, it discards the fragment.  On the other hand, if my fragment shader determines that the current fragment is near, but not in, the selected object, it needs to set the fragment depth to the depth sampled from the depth texture, because my big quad is at the wrong depth.

**Reply by romainguy** [reply]
Ah yeah that makes a lot of sense for this use case.
